smolyay/samples.py

```diff
@@ -0,0 +1,415 @@
+
+
+class NestedUnidimensionalPointSet(UnidimensionalPointSet):
+    """Set of unidimensional points assigned to levels.
+
+
+    Parameters
+    ----------
+    max_level ; int
+        the maximum level the points are used for
+    """
+
+    def __init__(self, domain, max_level):
+        super().__init__(domain)
+        self._max_level = None
+        self._num_points_per_level = None
+        self._start_level = None
+        self._end_level = None
+
+        self.max_level = max_level
+
+    @property
+    def max_level(self):
+        """int: maximum level to compute points for."""
+        return self._max_level
+
+    @max_level.setter
+    def max_level(self, value):
+        max_level = int(value)
+        if max_level != self._max_level:
+            self._max_level = max_level
+            self._valid_cache = False
+
+    @property
+    def num_points_per_level(self):
+        """numpy.ndarray: number of points per level."""
+        if not self._valid_cache:
+            self._create()
+            self._valid_cache = True
+        return self._num_points_per_level
+
+    @property
+    def start_level(self):
+        """numpy.ndarray: the starting index of each level."""
+        if not self._valid_cache:
+            self._create()
+            self._valid_cache = True
+        return self._start_level
+
+    @property
+    def end_level(self):
+        """numpy.ndarray: the ending index of each level."""
+        if not self._valid_cache:
+            self._create()
+            self._valid_cache = True
+        return self._end_level
+
+
+class ClenshawCurtisPointSet(UnidimensionalPointSet):
+    r"""Set of unidimensional points for Clenshaw Curtis sampling
+
+    The :attr:`points` for this interpolation scheme are the extrema of the
+    Chebyshev polynomials of the first kind on the domain :math:`[-1, 1]`:
+
+    .. math::
+
+        x_i^* = -\cos(\pi i/n), i = 0,...,n
+
+    For the special case :math:`n = 0`, there is only one point :math:`x_0^* = 0`.
+
+    Parameters
+    ----------
+    degree : int
+        degree of the Chebyshev polynomial of the first kind to get extrema from
+    """
+
+    def __init__(self, domain, degree):
+        super().__init__(domain)
+        self._degree = None
+
+        self.degree = degree
+
+    @property
+    def degree(self):
+        """int: degree of polynomial to create points for."""
+        return self._degree
+
+    @degree.setter
+    def degree(self, value):
+        degree = int(value)
+        if degree != self._degree:
+            self._degree = degree
+            self._valid_cache = False
+
+    def _create(self):
+        r"""Create the points in the set.
+
+        Generating the extrema of a Cheybshev polynomial of the first kind at
+        a given degree.
+        """
+        if self.degree > 0:
+            points = -numpy.cos(
+                numpy.pi * numpy.linspace(0, self.degree, self.degree + 1) / self.degree
+            )
+        else:
```

Comment:
I've tripped myself reading this code a few times because of the `>`. A comment here like this would help clarify:
```suggestion
        else:
            # special case where degree == 0
```